cornac/models/dnntsp/dnntsp.py

```diff
@@ -0,0 +1,639 @@
+            h = gcn(graph, h, edges_weight)
+            h = bn(h.transpose(1, -1)).transpose(1, -1)
+            h = relu(h)
+        return h
+
+
+class StackedWeightedGCNBlocks(nn.ModuleList):
+    def __init__(self, *args, **kwargs):
+        super(StackedWeightedGCNBlocks, self).__init__(*args, **kwargs)
+
+    def forward(self, *input):
+        g, nodes_feature, edge_weights = input
+        h = nodes_feature
+        for module in self:
+            h = module(g, h, edge_weights)
+        return h
+
+
+class TemporalSetPrediction(nn.Module):
+    def __init__(self, n_items, emb_dim):
+        super(TemporalSetPrediction, self).__init__()
+
+        self.embedding_matrix = nn.Embedding(n_items, emb_dim)
+
+        self.emb_dim = emb_dim
+        self.n_items = n_items
+        self.stacked_gcn = StackedWeightedGCNBlocks(
+            [WeightedGCN(emb_dim, [emb_dim], emb_dim)]
+        )
+
+        self.masked_self_attention = MaskedSelfAttention(
+            input_dim=emb_dim, output_dim=emb_dim
+        )
+
+        self.aggregate_nodes_temporal_feature = AggregateNodesTemporalFeature(
+            item_embed_dim=emb_dim
+        )
+
+        self.global_gated_update = GlobalGatedUpdate(
+            n_items=n_items, embedding_matrix=self.embedding_matrix
+        )
+
+        self.fc_output = nn.Linear(emb_dim, 1)
+
+    def forward(
+        self,
+        batch_graph,
+        batch_nodes_feature,
+        batch_edges_weight,
+        batch_lengths,
+        batch_nodes,
+    ):
+        # perform weighted gcn on dynamic graphs (n_1+n_2+..., T_max, item_embed_dim)
+        batch_nodes_output = [
+            self.stacked_gcn(graph, nodes_feature, edges_weight)
+            for graph, nodes_feature, edges_weight in zip(
+                batch_graph, batch_nodes_feature, batch_edges_weight
+            )
+        ]
+
+        # self-attention in time dimension, (n_1+n_2+..., T_max,  item_embed_dim)
+        batch_nodes_output = [
+            self.masked_self_attention(nodes_output)
+            for nodes_output in batch_nodes_output
+        ]
+
+        # aggregate node features in temporal dimension, (n_1+n_2+..., item_embed_dim)
+        batch_nodes_output = [
+            self.aggregate_nodes_temporal_feature(graph, lengths, nodes_output)
+            for graph, lengths, nodes_output in zip(
+                batch_graph, batch_lengths, batch_nodes_output
+            )
+        ]
+
+        # (batch_size, n_items, item_embed_dim)
+        batch_nodes_output = [
+            self.global_gated_update(graph, nodes, nodes_output)
+            for graph, nodes, nodes_output in zip(
+                batch_graph, batch_nodes, batch_nodes_output
+            )
+        ]
+
+        # (batch_size, n_items)
+        outputs = self.fc_output(torch.stack(batch_nodes_output)).squeeze()
+
+        return outputs
+
+
+def get_edges_weight(history_baskets):
+    edges_weight_dict = defaultdict(float)
+    for basket_items in history_baskets:
```

Comment:
How about using [itertools.combinations](https://docs.python.org/3/library/itertools.html#itertools.combinations) to simplify the two nested loops?